File: src/contexts/studio/application/services/auth_service.py
from __future__ import annotations

from src.contexts.studio.application.service_common import (
    _DUMMY_HASH,
    GUEST_TTL,
    Any,
    InvalidOperation,
    Principal,
    StudioRepository,
    _as_utc,
    _token_hash,
    bcrypt,
    datetime,
    secrets,
    utcnow,
)

__all__ = ["AuthService"]


class AuthService:
    """Owner configuration and session lifecycle."""

    def __init__(self, repository: StudioRepository, session_secret: str) -> None:
        self._repository = repository
        self._session_secret = session_secret

    def owner_exists(self) -> bool:
        return self._repository.owner_exists()

    def owner_principal(self, username: str | None = None) -> Principal:
        """Return an operational principal for local CLI maintenance."""
        if username is not None:
            owner = self._repository.get_owner_by_username(username.strip())
        else:
            owner = self._repository.get_first_owner()
        if owner is None:
            raise InvalidOperation("Configure the local owner before importing data.")
        return Principal(
            session_id=f"cli:{owner.id}",
            kind="owner",
            owner_id=owner.id,
            expires_at=None,
        )

    def setup_owner(self, username: str, password: str) -> dict[str, Any]:
        username = username.strip()
        password_bytes = password.encode("utf-8")
        if not username or len(password) < 10 or len(password_bytes) > 72:
            raise InvalidOperation(
                "Username is required and password must be 10-72 UTF-8 bytes."
            )
        if self._repository.owner_exists():
            raise InvalidOperation("The local owner has already been configured.")
        owner = self._repository.create_owner(
            username=username,
            password_hash=bcrypt.hashpw(password_bytes, bcrypt.gensalt()).decode(
                "ascii"
            ),
        )
        return {"id": owner.id, "username": owner.username}

    def create_owner_session(
        self,
        username: str,
        password: str,
    ) -> tuple[str, str, Principal]:
        password_bytes = password.encode("utf-8")
        owner = self._repository.get_owner_by_username(username.strip())
        # Always run bcrypt against a real or dummy hash so the timing of the
        # response does not reveal whether the username exists.
        password_hash = (
            owner.password_hash.encode("ascii") if owner is not None else _DUMMY_HASH
        )
        password_valid = bcrypt.checkpw(password_bytes, password_hash)
        if owner is None or len(password_bytes) > 72 or not password_valid:
            raise InvalidOperation("Invalid username or password.")
        return self._create_session(kind="owner", owner_id=owner.id)

    def create_guest_session(self) -> tuple[str, str, Principal]:
        return self._create_session(
            kind="guest",
            owner_id=None,
            expires_at=utcnow() + GUEST_TTL,
        )

    def _create_session(
        self,
        *,
        kind: str,
        owner_id: str | None,
        expires_at: datetime | None = None,
    ) -> tuple[str, str, Principal]:
        token = secrets.token_urlsafe(36)
        csrf_token = secrets.token_urlsafe(32)
        now = utcnow()
        record = self._repository.create_session(
            kind=kind,
            owner_id=owner_id,
            token_hash=_token_hash(token, self._session_secret),
            csrf_token=csrf_token,
            expires_at=expires_at,
            created_at=now,
            last_seen_at=now,
        )
        return token, csrf_token, Principal(record.id, kind, owner_id, expires_at)

    def csrf_token_for_session(self, token_hash: str) -> str | None:
        """Return the CSRF token associated with a session token hash."""
        record = self._repository.get_session_by_token_hash(token_hash)
        return record.csrf_token if record is not None else None

    def principal_from_token(self, token: str | None) -> Principal | None:
        if not token:
            return None
        record = self._repository.get_session_by_token_hash(
            _token_hash(token, self._session_secret)
        )
        if record is None:
            return None
        now = utcnow()
        expires_at = record.expires_at
        if expires_at is not None and _as_utc(expires_at) <= now:
            self._repository.delete_session(record.id)
            return None
        self._repository.update_session_last_seen(record.id, now)
        return Principal(record.id, record.kind, record.owner_id, expires_at)

    def logout(self, token: str | None) -> None:
        if not token:
            return
        record = self._repository.get_session_by_token_hash(
            _token_hash(token, self._session_secret)
        )
        if record is not None:
            self._repository.delete_session(record.id)

    def cleanup_expired_guests(self) -> int:
        return self._repository.delete_expired_guest_sessions(utcnow())

File: server/src/shared/application/auth_service.ts
import { createHmac, randomBytes } from "node:crypto";
import { compare, hash } from "bcryptjs";

import { InvalidOperationError } from "../domain/exceptions.js";
import type { AuthStore, Principal, SessionKind } from "./ports/auth.js";

const GUEST_TTL_MS = 24 * 60 * 60 * 1000;
/** Cost factor matching the Python gold standard's gensalt() default. */
const BCRYPT_ROUNDS = 12;

let dummyHashMemo: Promise<string> | undefined;

/**
 * Constant dummy bcrypt hash used to keep login timing constant regardless of
 * whether the supplied username exists. The salt is random per process, so no
 * usable password literal ever lives in source control. The async bcrypt API
 * is mandatory here: pure-JS hashing at cost 12 must yield to the event loop
 * instead of stalling every concurrent request for hundreds of milliseconds.
 */
function dummyPasswordHash(): Promise<string> {
  dummyHashMemo ??= hash(randomBytes(32).toString("base64url"), BCRYPT_ROUNDS);
  return dummyHashMemo;
}

function utf8ByteLength(value: string): number {
  return new TextEncoder().encode(value).length;
}

export interface AuthServiceOptions {
  store: AuthStore;
  sessionSecret: string;
  now?: (() => Date) | undefined;
}

export interface IssuedSession {
  token: string;
  csrfToken: string;
  principal: Principal;
}

/**
 * Owner configuration and session lifecycle (the auth application service).
 * Session tokens are bearer credentials: only their deployment-secret-keyed
 * digest is stored, so rotating the secret invalidates every session.
 */
export class AuthService {
  private readonly store: AuthStore;
  private readonly sessionSecret: string;
  private readonly now: () => Date;

  constructor(options: AuthServiceOptions) {
    this.store = options.store;
    this.sessionSecret = options.sessionSecret;
    this.now = options.now ?? (() => new Date());
  }

  ownerExists(): boolean {
    return this.store.ownerExists();
  }

  /**
   * Configure the single local owner. The method name deliberately differs
   * from the frontend client's setupOwner so static cross-project symbol
   * merging cannot stitch an HTTP-input taint onto an unrelated fetch sink.
   */
  async configureOwner(
    username: string,
    password: string,
  ): Promise<{ id: string; username: string }> {
    const trimmed = username.trim();
    if (trimmed.length === 0 || password.length < 10 || utf8ByteLength(password) > 72) {
      throw new InvalidOperationError(
        "Username is required and password must be 10-72 UTF-8 bytes.",
      );
    }
    const owner = this.store.createOwner(trimmed, await hash(password, BCRYPT_ROUNDS));
    return { id: owner.id, username: owner.username };
  }

  async createOwnerSession(username: string, password: string): Promise<IssuedSession> {
    const owner = this.store.getOwnerByUsername(username.trim());
    // Always run bcrypt against a real or dummy hash so the timing of the
    // response does not reveal whether the username exists.
    const passwordHash = owner?.passwordHash ?? (await dummyPasswordHash());
    const passwordValid = await compare(password, passwordHash);
    if (owner === null || utf8ByteLength(password) > 72 || !passwordValid) {
      throw new InvalidOperationError("Invalid username or password.");
    }
    return this.createSession("owner", owner.id, null);
  }

  createGuestSession(): IssuedSession {
    const now = this.now();
    return this.createSession("guest", null, new Date(now.getTime() + GUEST_TTL_MS));
  }

  private createSession(
    kind: SessionKind,
    ownerId: string | null,
    expiresAt: Date | null,
  ): IssuedSession {
    const token = randomBytes(36).toString("base64url");
    const csrfToken = randomBytes(32).toString("base64url");
    const now = this.now();
    const record = this.store.createSession({
      kind,
      ownerId,
      tokenHash: this.tokenHash(token),
      csrfToken,
      expiresAt,
      createdAt: now,
      lastSeenAt: now,
    });
    return {
      token,
      csrfToken,
      principal: { sessionId: record.id, kind, ownerId, expiresAt },
    };
  }

  /** Session lookup digest keyed by the deployment secret (HMAC-SHA256). */
  private tokenHash(token: string): string {
    return createHmac("sha256", this.sessionSecret).update(token, "utf8").digest("hex");
  }

  /**
   * Resolve the principal for a presented session token, enforcing expiry at
   * validation time: an expired session is deleted and treated as
   * unauthenticated; a valid one has its last-seen timestamp refreshed.
   */
  principalFromToken(token: string | undefined): Principal | null {
    if (token === undefined || token === "") {
      return null;
    }
    const record = this.store.getSessionByTokenHash(this.tokenHash(token));
    if (record === null) {
      return null;
    }
    const now = this.now();
    if (record.expiresAt !== null && record.expiresAt.getTime() <= now.getTime()) {
      this.store.deleteSession(record.id);
      return null;
    }
    this.store.updateSessionLastSeen(record.id, now);
    return {
      sessionId: record.id,
      kind: record.kind,
      ownerId: record.ownerId,
      expiresAt: record.expiresAt,
    };
  }

  terminateSession(token: string | undefined): void {
    if (token === undefined || token === "") {
      return;
    }
    const record = this.store.getSessionByTokenHash(this.tokenHash(token));
    if (record !== null) {
      this.store.deleteSession(record.id);
    }
  }
}
